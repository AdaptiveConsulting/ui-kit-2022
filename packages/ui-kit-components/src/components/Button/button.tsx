import { Button, SvgIcon } from '@mui/material';
import * as React from 'react';

import { ArrowLeft, ArrowRight } from '../../icons';

export type VariantType = 'PRIMARY' | 'SECONDARY' | 'TERTIARY' | undefined;
export interface Props {
  children?: JSX.Element | string;
  variant?: VariantType;
  disabled?: boolean;
  icon?: 'left' | 'right';
  click?: (e: React.BaseSyntheticEvent) => void;
}

const ButtonDefault: React.FC<Props> = ({
  variant = 'PRIMARY',
  disabled = false,
  children = '',
  icon,
  click,
}) => {
  if (icon === 'left') {
    return (
      <Button
        disableRipple
        variant={variant}
        disabled={disabled}
        onClick={click}
        startIcon={
          <SvgIcon
            component={ArrowLeft}
            viewBox="-3 0 16 16"
            sx={{ width: '16px', height: '16px' }}
          />
        }
        sx={{ '& .MuiButton-startIcon': { mr: '4px' } }}
      >
        {children}
      </Button>
    );
  } else if (icon === 'right') {
    return (
      <Button
        disableRipple
        variant={variant}
        disabled={disabled}
        onClick={click}
        endIcon={
          <SvgIcon
            component={ArrowRight}
            viewBox="3 0 16 16"
            sx={{ width: '16px', height: '16px' }}
          />
        }
        sx={{ '& .MuiButton-endIcon': { ml: '4px' } }}
      >
        {children}
      </Button>
    );
  }

  return (
    <Button disableRipple variant={variant} disabled={disabled} onClick={click}>
      {children}
    </Button>
  );
};

export default ButtonDefault;
