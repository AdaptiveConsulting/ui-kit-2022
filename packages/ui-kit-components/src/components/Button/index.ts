import Button, { Props as ButtonProps, VariantType as ButtonVariant } from './button';

declare module '@mui/material/Button' {
  interface ButtonPropsVariantOverrides {
    PRIMARY: true;
    SECONDARY: true;
    TERTIARY: true;
    outlined: false;
    text: false;
    contained: false;
  }
}

export { Button };
export type { ButtonProps, ButtonVariant };
