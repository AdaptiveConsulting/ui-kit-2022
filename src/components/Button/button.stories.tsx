import { ComponentMeta, ComponentStory } from '@storybook/react';
import * as React from 'react';

import ButtonDefault from './button';

export default {
  title: 'Components/Button',
  component: ButtonDefault,
  argTypes: {
    variant: {
      options: ['PRIMARY', 'SECONDARY', 'TERTIARY'],
      control: { type: 'radio' },
    },
  },
} as ComponentMeta<typeof ButtonDefault>;

const Template: ComponentStory<typeof ButtonDefault> = (args) => (
  <ButtonDefault {...args}></ButtonDefault>
);

export const Basic = Template.bind({});

Basic.args = {
  variant: 'PRIMARY',
  children: 'Hello Adaptive',
  click: () => alert('This is a text button'),
};

export const LeftIcon = Template.bind({});

LeftIcon.args = {
  variant: 'PRIMARY',
  children: 'Hello Adaptive',
  icon: 'left',
  click: () => alert('This is a text button'),
};

export const RightIcon = Template.bind({});

RightIcon.args = {
  variant: 'PRIMARY',
  children: 'Hello Adaptive',
  icon: 'right',
  click: () => alert('This is a text button'),
};
