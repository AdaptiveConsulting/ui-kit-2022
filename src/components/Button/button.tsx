import { Button, SvgIcon } from '@mui/material';
import * as React from 'react';

import { ArrowLeft, ArrowRight } from '../../icons';
type VariantType = 'PRIMARY' | 'SECONDARY' | 'TERTIARY';

interface Props {
  children?: JSX.Element;
  variant?: VariantType;
  disabled?: boolean;
  icon: 'left' | 'right' | undefined;
  click: () => void;
}

const ButtonDefault: React.FC<Props> = ({
  variant = 'PRIMARY',
  disabled = false,
  children = '',
  icon,
  click,
}) => {
  if (icon === 'left') {
    return (
      <Button
        disableRipple
        variant={variant}
        disabled={disabled}
        onClick={click}
        startIcon={
          <SvgIcon
            component={ArrowLeft}
            viewBox="-3 0 16 16"
            sx={{ width: '16px', height: '16px' }}
          />
        }
        sx={{ '& .MuiButton-startIcon': { mr: '4px' } }}
      >
        {children}
      </Button>
    );
  } else if (icon === 'right') {
    return (
      <Button
        disableRipple
        variant={variant}
        disabled={disabled}
        onClick={click}
        endIcon={
          <SvgIcon
            component={ArrowRight}
            viewBox="3 0 16 16"
            sx={{ width: '16px', height: '16px' }}
          />
        }
        sx={{ '& .MuiButton-endIcon': { ml: '4px' } }}
      >
        {children}
      </Button>
    );
  }

  return (
    <Button disableRipple variant={variant} disabled={disabled} onClick={click}>
      {children}
    </Button>
  );
};

export default ButtonDefault;
