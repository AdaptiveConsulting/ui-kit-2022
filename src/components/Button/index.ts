declare module '@mui/material/Button' {
  interface ButtonPropsVariantOverrides {
    PRIMARY: true;
    SECONDARY: true;
    TERTIARY: true;
    outlined: false;
    text: false;
    contained: false;
  }
}

export {};
