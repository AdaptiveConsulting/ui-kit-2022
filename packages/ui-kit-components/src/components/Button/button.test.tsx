import '@testing-library/jest-dom';

import { act, fireEvent, render, screen } from '@testing-library/react';

import Button from './button';

describe('Button component tests', () => {
  describe('Given render a simple button', () => {
    const fn = jest.fn(() => {});
    beforeEach(() => {
      render(<Button click={fn}>Test Button</Button>);
    });

    describe('When do the snapshots test', () => {
      it('Then should be the as same the snapshot', () => {
        expect(screen.getAllByText('Test Button')[0]).toMatchSnapshot();
      });
    });

    describe('When it is a simple button', () => {
      it('Then should have the button label `Test Button`', () => {
        expect(screen.getAllByText('Test Button')[0]).toBeInTheDocument();
      });
    });

    describe('When the button has been clicked', () => {
      beforeEach(() => {
        fireEvent.click(screen.getByText('Test Button'));
      });
      it('Then the onClick function should be fired once', () => {
        expect(fn).toBeCalled();
      });
    });
  });

  describe('Given render a button with/without icon', () => {
    describe('When it is a simple button', () => {
      beforeEach(() => {
        render(<Button>Normal Button</Button>);
      });

      it('Then should get a normal button without icon', () => {
        expect(screen.getByTestId('normal')).toBeInTheDocument();
      });
    });

    describe('When it is a button with left icon', () => {
      beforeEach(() => {
        render(<Button icon="left">Left Button</Button>);
      });

      it('Then should get a button with left icon', () => {
        expect(screen.getByTestId('left-icon')).toBeInTheDocument();
      });
    });

    describe('When it is a button with right icon', () => {
      beforeEach(() => {
        render(<Button icon="right">Right Button</Button>);
      });

      it('Then should get a button with left icon', () => {
        expect(screen.getByTestId('right-icon')).toBeInTheDocument();
      });
    });
  });

  describe('Given a simple button', () => {
    beforeEach(() => {
      render(<Button>Focus Button</Button>);
    });

    describe('When focus the button', () => {
      beforeEach(() => {
        const element = screen.getByTestId('normal');
        act(() => {
          element.focus();
        });
      });

      it('Then should have the Mui-focusVisible classname', () => {
        const element = screen.getByTestId('normal');
        expect(element).toHaveClass('Mui-focusVisible');
      });

      describe('When the button loses focus', () => {
        beforeEach(() => {
          const element = screen.getByTestId('normal');
          act(() => {
            element.blur();
          });
        });

        it('Then should not have the Mui-focusVisible classname', () => {
          const element = screen.getByTestId('normal');
          expect(element).not.toHaveClass('Mui-focusVisible');
        });
      });
    });
  });
});
