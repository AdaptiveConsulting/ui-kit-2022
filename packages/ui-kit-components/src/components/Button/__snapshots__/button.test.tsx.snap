// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component tests Given render a simple button When do the snapshots test Then should be the as same the snapshot 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-PRIMARY MuiButton-PRIMARYPrimary MuiButton-sizeMedium MuiButton-PRIMARYSizeMedium MuiButton-root MuiButton-PRIMARY MuiButton-PRIMARYPrimary MuiButton-sizeMedium MuiButton-PRIMARYSizeMedium css-1w1rijm-MuiButtonBase-root-MuiButton-root"
  data-testid="normal"
  tabindex="0"
  type="button"
>
  Test Button
</button>
`;
